waves-integration/src/main/scala/com/wavesplatform/dex/grpc/integration/clients/combined/CombinedStreamActor.scala

```diff
@@ -0,0 +1,281 @@
+import monix.execution.Scheduler
+import monix.reactive.subjects.ConcurrentSubject
+import org.slf4j.event.Level
+
+import java.util.concurrent.atomic.AtomicInteger
+
+object CombinedStreamActor {
+  sealed trait Message extends Product with Serializable
+
+  sealed trait Command extends Message
+
+  object Command {
+    case class Start(fromHeight: Int) extends Command
+
+    case class UpdateProcessedHeight(height: Int) extends Command
+
+    case object Restart extends Command
+
+    case object Continue extends Command
+
+    case class ProcessUtxSystemEvent(evt: SystemEvent) extends Command
+    case class ProcessBlockchainUpdatesSystemEvent(evt: SystemEvent) extends Command
+
+    case class ProcessUtxEvent(evt: UtxEvent) extends Command
+    case class ProcessBlockchainUpdatesEvent(evt: SubscribeEvent) extends Command
+  }
+
+  def apply(
+    settings: Settings,
+    processedHeight: AtomicInteger,
+    blockchainUpdates: BlockchainUpdatesControlledStream,
+    utxEvents: UtxEventsControlledStream,
+    status: ConcurrentSubject[Status, Status],
+    output: ConcurrentSubject[WavesNodeEvent, WavesNodeEvent]
+  )(implicit monixScheduler: Scheduler): Behavior[Command] = Behaviors.setup[Command] { context =>
+    // Subscribe on events
+
+    blockchainUpdates.systemStream.foreach(context.self ! Command.ProcessBlockchainUpdatesSystemEvent(_))
+    blockchainUpdates.stream.foreach(context.self ! Command.ProcessBlockchainUpdatesEvent(_))
+
+    utxEvents.systemStream.foreach(context.self ! Command.ProcessUtxSystemEvent(_))
+    utxEvents.stream.foreach(context.self ! Command.ProcessUtxEvent(_))
+
+    // Utilities
+
+    val ignore: Command => Behavior[Command] = _ => Behaviors.same
+    def partial(f: PartialFunction[Command, Behavior[Command]]): PartialFunction[Command, Behavior[Command]] = f
+
+    def logAndIgnore(text: String): Behavior[Command] = {
+      context.log.warn(text)
+      Behaviors.same[Command]
+    }
+
+    // Possible transitions:
+    //   initial -> starting
+    //   starting -> working, quiet, stopping, closing
+    //   working -> stopping, closing
+    //   quiet -> starting, closing
+    //   stopping -> quiet, closing
+    //   closing -> closed
+
+    def closed: Behavior[Command] = {
+      context.log.info("closed")
+      status.onNext(Status.Closing(blockchainUpdates = true, utxEvents = true))
+      Behaviors.receiveMessage[Command](x => logAndIgnore(s"Unexpected $x"))
+    }
+
+    def closing(utxEventsClosed: Boolean, blockchainUpdatesClosed: Boolean): Behavior[Command] = {
+      context.log.info(s"closing(utx=$utxEventsClosed, bu=$blockchainUpdatesClosed)")
+      status.onNext(Status.Closing(blockchainUpdates = blockchainUpdatesClosed, utxEvents = utxEventsClosed))
+      Behaviors.receiveMessage[Command] {
+        case Command.ProcessUtxSystemEvent(SystemEvent.Closed) =>
+          if (utxEventsClosed) logAndIgnore("utx has already closed")
```

Comment:
> possible missed handling

Actually we should do only logging in the case of `utxEventsClosed && !blockchainUpdatesClosed`, because `utxEventsClosed=true` means that we have already received this event and called `blockchainUpdates.close()`.

Added the documentation.